### PR TITLE
Use native AAs for Mangler.types and idents

### PR DESCRIPTION
**Pull request by ibuclaw**
Continuation of #5428.

### CODE REVIEW

File: src/dmd/dmangle.d [hunk 1]
```diff
@@ -225,13 +223,13 @@ public:
     {
         if (!t.isTypeBasic())
         {
-            auto p = cast(size_t*)dmd_aaGet(&types, cast(Key)t);
-            if (*p)
+            auto p = cast(void*)t in types;
+            if (p)
```
Comment (ibuclaw) [inline src/dmd/dmangle.d]: Blasts (to put it nicely).  I've just realised why its failing. (check should be `p && *p` :-)

**Review by JinShil — APPROVED** [review]

File: src/dmd/dmangle.d [hunk 2]
```diff
@@ -170,9 +169,8 @@ private extern (C++) final class Mangler : Visitor
 {
     alias visit = Visitor.visit;
 public:
-    static assert(Key.sizeof == size_t.sizeof);
-    AA* types;
-    AA* idents;
+    size_t[void*] types;
+    size_t[void*] idents;
```
Comment (jacob-carlborg) [inline src/dmd/dmangle.d]: Why is `void*` used instead of the actual types inserted in the AA?

Comment (jacob-carlborg) [reply src/dmd/dmangle.d]: Found the answer in another PR.

**Review by jacob-carlborg — APPROVED** [review]

File: src/dmd/dmangle.d (hunk 2 above)
Comment (ibuclaw) [reply src/dmd/dmangle.d]: Also Walter explained in the linked PR.

Comment (ibuclaw) [reply src/dmd/dmangle.d]: That it is an optimization is not too obvious unless you know. ☺️

Comment (rainers) [reply src/dmd/dmangle.d]: > That it is an optimization is not too obvious unless you know.

That's not just an optimization, but a necessity as using C++ classes as key just crashes when trying to access the type info.

File: src/dmd/dmangle.d [hunk 3]
```diff
@@ -225,13 +223,13 @@ public:
...
-            auto p = cast(size_t*)dmd_aaGet(&types, cast(Key)t);
-            if (*p)
+            auto p = cast(void*)t in types;
+            if (p && *p)
             {
                 writeBackRef(buf.offset - *p);
                 return true;
             }
-            *p = buf.offset;
+            types[cast(void*)t] = buf.offset;
```
Comment (rainers) [inline src/dmd/dmangle.d]: Not too happy about the duplicate lookups here...

File: src/dmd/dmangle.d (hunk 2 above)
Comment (jacob-carlborg) [reply src/dmd/dmangle.d]: Perhaps that should be an error instead?

Comment (rainers) [reply src/dmd/dmangle.d]: >Perhaps that should be an error instead?

Or just fall back to void* typeinfo. See also https://issues.dlang.org/show_bug.cgi?id=13875 and https://issues.dlang.org/show_bug.cgi?id=15813.

File: src/dmd/dmangle.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/dmangle.d]: Me neither. I suspect it can cause serious slowdowns.

File: src/dmd/dmangle.d [hunk 4]
```diff
@@ -251,13 +249,13 @@ public:
...
-        auto p = cast(size_t*)dmd_aaGet(&idents, cast(Key)id);
-        if (*p)
+        auto p = cast(void*)id in idents;
+        if (p && *p)
         {
             writeBackRef(buf.offset - *p);
             return true;
         }
-        *p = buf.offset;
+        idents[cast(void*)id] = buf.offset;
```
Comment (WalterBright) [inline src/dmd/dmangle.d]: More duplicate lookups.

**Comment by WalterBright** [comment]
May need to revert this due to slowdowns caused by duplicate lookups.

**Comment by ibuclaw** [comment]
@WalterBright - maybe you can think of an alternative?  You introduced duplicate lookups in #5428 too.

**Comment by WalterBright** [comment]
This was reverted and added to the Phantom Zone because it is a good idea, but awaits a resolution of the double lookup problem which affects most of the uses of built-in arrays in the dmd front end.

**Comment by ibuclaw** [comment]
@WalterBright could root.aav perhaps be converted into a thin wrapper around druntime AAs in the meantime perhaps? 

The one difference would be to use aaGet instead of aaIn for the overload of operator `in`.

**Pull request merged**